### DISCUSSION TITLE
Maximum call stack size exceeded on next 11

### DISCUSSION THREAD

**Discussion by joshma** [vercel/next.js#30785] [Help]
### What version of Next.js are you using?

11.1.2

### What version of Node.js are you using?

16.13.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

When running `next dev`, all requests error with `Maximum call stack size exceeded`:

```
$ next dev
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Loaded env from /<redacted>/.env.local
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

info  - Using external babel configuration from /<redacted>/.babelrc
error - static/chunks/pages/_app.js
Maximum call stack size exceeded
event - build page: /next/dist/pages/_error
wait  - compiling...
error - static/chunks/pages/_app.js
Maximum call stack size exceeded
wait  - compiling...
error - static/chunks/pages/_app.js
Maximum call stack size exceeded
event - build page: /next/dist/pages/_error
wait  - compiling...
error - static/chunks/pages/_app.js
Maximum call stack size exceeded
```

I don't see any other logs, so I'm not entirely sure if this is a Next.js issue or some dependency. Running `rm -rf .next/ node_modules/` and re-installing doesn't work either, but _restarting my laptop_ does.

Is there some way I can get the full stack?

This has been fine for a few months now and only started happening frequently (about once a day) in the last 2 weeks.

### Expected Behavior

I'd expect the page to load normally.

### To Reproduce

I can't repro this unfortunately, so I'm looking for troubleshooting steps to further isolate the error.

**Comment by joshma** [comment]
Alright, I'm now convinced this is coming from some part of webpack (just based on Googling and finding other issues from the past). But setting `--stack-size=10000` in the `next` shebang, just to test, did work - so I don't think this is some recursive loop but just a genuine stack size problem.

Feel free to close if this is out of scope of next.js, but I figured it might be relevant if next.js is fiddling with stack size somewhere, managing the webpack config, or has some flag I can use to make debugging this better.

**Reply by shafkathullah** [reply]
I am having this issue on my M1 mac, also I am using tailwindcss with postcss as well, where is this next shebang so to add manully add stack size?

**Comment by timneutkens** [comment]
You can use the normal debugging methods of profiling during the development run and checking the memory usage. Next.js 12 has many improvements including improved memory management so I'd recommend to start by upgrading to the latest version 👍

**Reply by joshma** [reply]
While I get this is hard to investigate, it's also hard for us too. :) So I'll need some pointers here.

I don't believe this is a memory management issue, although I guess you could consider managing stack size to be partly memory related. Also, we've noticed `yarn build` _sometimes_ helps, and a reboot _sometimes_ helps, both of which are surprising. The only folks on my team experiencing this are on M1 MBPs, the others are on Linux, so that's one other thing to note.

**Comment by timneutkens** [comment]
Converted to discussions given that there is no reproduction that can be investigated so it's not actionable to investigate from our side 👍

**Comment by JackCA** [comment]
We're also experiencing this on the latest version of next.js and every build tool we use -- is it possibly related to node v16?

**Reply by joshma** [reply]
Hm maybe! I had to switch to node 16 recently, was on node 15 earlier.

Note that I don't think this is a memory issue, it's just a deep stack trace.

**Reply by timneutkens** [reply]
@JackCA maybe your issue is #30330, can you make sure to try `12.0.2` 🙏

**Reply by JackCA** [reply]
thanks for the heads up, but unfortunately, we were on 12.0.2 when we were getting the issue. I can confirm that downgrading to v14 has fixed it

**Reply by OwenMelbz** [reply]
I'm on v17.0.1 and experiencing this as well..

**Reply by lucascaton** [reply]
Downgrading Node.js from version `16.13.0` to `14.18.1` seems to have solved the issue on my project 👍

**Reply by stevenxu-db** [reply]
We saw the same symptoms where the error would present only on Node 16 and work fine on Node 14. This was traced down to the regexp match check with pattern `/[^\n]+\n?|\n/g` in the vendored webpack within [Next 11.1.2](https://github.com/vercel/next.js/blob/v11.1.2/packages/next/compiled/webpack-sources3/index.js) running over a string that included a **~2^22-length encoded sourcemap that didn't include a newline**.

With the Node debugger paused at this point, passing in the first 4,194,285 characters of the offending string would consistently trip `RangeError: Maximum call stack size exceeded` and passing in the first 4,194,284 would work consistently very quickly. Our impression is that something in V8 internals tries to allocate stack memory in some integer proportion to the length of the repeating sequence in the regexp and that this is hitting some V8 check in [RegExpMatchFast](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/builtins/regexp-match.tq;l=135?q=RegExpMatchFast&ss=chromium) and causing a fatal error in V8.

The bug was not reproducible using the same string and pattern in a REPL, so some environmental factor in the Node process that Next spawns to handle HMR has a different configuration.

We ultimately worked around the issue by reworking our infrastructure to avoid emitting that source map.

**Comment by blakewilson** [comment]
I'm seeing the same error on `next@11.1.2`.

```
error - static/chunks/pages/_app.js
Maximum call stack size exceeded
```

Node version: v16.13.0

As others have noted, downgraded to Node v14 does indeed resolve the issue.

This is really hard to reproduce on node v16, I believe it's in conjunction with other packages, like referenced in https://github.com/vercel/next.js/issues/30330. For instance, I'm using `tailwind`.  Removing `tailwindcss` from PostCSS fixes the stack overflow. This seems to be happening at a low-level Webpack bundling step.

**Comment by elado** [comment]
Unclear if it's related but I've seen `Maximum call stack size exceeded` from code inside Next's bundled Webpack in a production build. The faulty line was:

```
s.match(/[^\n]+\n?|\n/g);
```

on a big source file that was also included in `next-transpile-modules`. I managed to bypass it by excluding it from `next-transpile-modules` but not sure what the root cause is, or why Node (14) would fail even if the string is large.

**Reply by blakewilson** [reply]
I have a ~500kb/15k line file for a GraphQL schema represented in TS. I wonder if this is related 👀 I'll test and report back

**Comment by lucascaton** [comment]
I'm still having issues with the latest versions, ie.: `Node.js 16.14.0` and `Next 12.0.10`:

```js
wait  - compiling /_error (client and server)...
error - ./node_modules/next/dist/build/webpack/loaders/css-loader/src/index.js??ruleSet[1].rules[2].oneOf[5].use[1]!./node_modules/next/dist/build/webpack/loaders/postcss-loader/src/index.js??ruleSet[1].rules[2].oneOf[5].use[2]!./node_modules/next/dist/build/webpack/loaders/resolve-url-loader/index.js??ruleSet[1].rules[2].oneOf[5].use[3]!./node_modules/next/dist/compiled/sass-loader/cjs.js??ruleSet[1].rules[2].oneOf[5].use[4]!./src/pages/path/myPage.module.scss
Cannot read properties of undefined (reading 'indexOf')
```

**Comment by catfish233** [comment]
I tried switching the version of the node from 16.16.0 to 16.0.0 via NVM.  Removed '.next' then 'yarn dev' again. It worked.

**Comment by pilotso11** [comment]
I'm having a similar problem, npm run dev is working but npm run build fails with the call stack error. 
RangeError: Maximum call stack size exceeded
    at getContextualType (... /node_modules/typescript/lib/typescript.js:71644:58)
Switching the node version had no effect for me.  (I tried 16.16.0, 16.0.0, 14.7.1 and 12.4.0).

Rolling back to an older version of typescript, 4.5.5, seems to have resolved it for me.   I just updated it in the package.json and ran npm install.   This works with both node 16.16.0 and 16.0.0.   

It works on typescript up to 4.5.5 (also tested 4.3.2 and 4.4.4) and fails on 4.6.4 and 4.7.4.

**Comment by Alireza17224** [comment]
It is likely that a function with an infinite loop is being called

**Comment by larsqa** [comment]
Agree with @Alireza17224, this is either a very deep (nasty) nested function or very likely a function that ends up calling itself.
You can get more information by temporarily increasing the callstack size with

```sh
node --stack-size=1200 ./node_modules/next/dist/cli/next-dev.js
# or
node --stack-size=1200 ./node_modules/next/dist/cli/next-start.js
```
This runs `npm run dev` or `npm run start` with an increased stack size. This command does not affect the default stack size.

**Comment by Irrelon** [comment]
Just trying to add to the potential debugging if I can. I was able to able to hit this issue as well. My steps were that I switched to using a custom .babelrc. When I did this, I got:

```
Failed to get source map: RangeError: Maximum call stack size exceeded
    at RegExp.exec (<anonymous>)
    at getSourceMapUrl (/Users/robevans/Development/Irrelon/irrelon-merchant/node_modules/next/dist/compiled/@next/react-dev-overlay/dist/middleware.js:1:41886)
    at getRawSourceMap (/Users/robevans/Development/Irrelon/irrelon-merchant/node_modules/next/dist/compiled/@next/react-dev-overlay/dist/middleware.js:1:41282)
    at /Users/robevans/Development/Irrelon/irrelon-merchant/node_modules/next/dist/compiled/@next/react-dev-overlay/dist/middleware.js:1:64341
    at step (/Users/robevans/Development/Irrelon/irrelon-merchant/node_modules/next/dist/compiled/@next/react-dev-overlay/dist/middleware.js:1:60403)
    at Object.next (/Users/robevans/Development/Irrelon/irrelon-merchant/node_modules/next/dist/compiled/@next/react-dev-overlay/dist/middleware.js:1:59688)
    at fulfilled (/Users/robevans/Development/Irrelon/irrelon-merchant/node_modules/next/dist/compiled/@next/react-dev-overlay/dist/middleware.js:1:59187)
```

My babelrc is very basic:

```
{
  "presets": ["next/babel"],
  "plugins": []
}
```

I'm using Next.js 13.0.1.

The failing regular expression is:

<img width="959" alt="image" src="https://user-images.githubusercontent.com/492293/203092269-bdf4dcac-a552-49c4-b362-f358e90f503d.png">

I'm seeing this intermittently. I don't know if it is related to the original issue or not but wanted to add the info in case this helps.

**Comment by demogoran** [comment]
In my case middleware.ts seems to be a reason.
If I remove it or put config with incorrect pathes, build passed.

**Comment by larsqa** [comment]
A second time I faced this issue.
This time, I imported to my `src/middleware.js` a helper function, which had `util.promisify(crypto.scrypt)`. This, in return, created a `Error [RangeError]: Maximum call stack size exceeded` error. Using it outside the middleware doesn't produce any issue.

### Edit
Made some changes and faced this error.
```
Error: The edge runtime does not support Node.js 'crypto' module.
```
Well - seems like I have to roll back to bcrypt instead, since their suggestion to use Web Crypto Api doesn't provide `scrypt`... 🙃

**Comment by Zeusxl60** [comment]
In my case, importing Head from next/document was what started the issue, but you also can't build while importing it from next/head on next build, so I imported it from next/head in _app.js.

**Comment by gabrieltira94** [comment]
I've just stumbled upon this issue today, on `14.2.6`.

What I did was: I created a function in an `utility file` and called it on both `client pages` and `API routes`.

The fix: I removed the import from the API routes. It was interfering with the middleware in my case.

**Comment by ramses44-17** [comment]
I have the same problem with nextjs v15 and nodejs v20.14.0.

**Reply by Mrahmani71** [reply]
I had the same problem, and after checking my code, I found the issue was in this component:

```tsx
// import { HomeLayout } from '@/components/layouts/home-layout'
import React from 'react'

interface LayoutProps {
  children: React.ReactNode
}

const HomeLayout = ({ children }: LayoutProps) => {
  return (
    <HomeLayout>
      {children}
    </HomeLayout>
  )
}

export default HomeLayout
```

If you spotted the problem: I was using a component inside itself with the same name, and I didn’t realize it. The IDE didn’t show any warning either. That’s why I got the "Maximum call stack size exceeded" error.

**Comment by MiroslavPetrik** [comment]
I am updating version from 15 to 16, and got this issue

```
D:\my-app\node_modules\next\dist\compiled\next-server\app-page-turbo.runtime.dev.js: Invalid source map. Only conformant source maps can be used to find the original code. Cause: Error: sourceMapURL could not be parsed
RangeError: Maximum call stack size exceeded
    at ignore-listed frames
```

v15 app (with turbopack) worked just fine

**Reply by MiroslavPetrik** [reply]
After some debugging, it seems like the issue happens while redirecting:
```ts
  if (await authorizedSession()) {
    redirect("/dashboard");
  }
```
Like, each time a new page is visited, which previously was not, I get the error.
Then on next attempt it works. But the behavior seems inconsistent.

I'm on Windows 11
